### PR TITLE
Add new NRT cache type DiscardReservedNodes

### PR DESCRIPTION
**Pull request by PiotrProkop**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Add new NRT cache type DiscardReservedNodes, which discards nodes from being scheduled if there are any pods reserved by NRT plugin for it.

DiscardReserved is intended to solve similiar problem as Overreserve Cache,
which is to minimize amount of incorrect scheduling decisions based on stale NRT data.
Unfortunately Overreserve cache only works for single-numa-node Topology Manager policy.
Dis tries to minimize amount of Admission Errors and non-optimal placement
when NodeResourceTopologyMatch plugin is used to schedule PODs requesting resources from multiple NUMA domains.
There are scenarios where using DiscardReserved won't mitigate drawbacks of using Passthrough cache.
NRT update is expected once PostBind triggers, but there's no guarantee about when this will happen.
In cases like:
- NFD(or any other component that advertises NRT) can be nonfunctional
- network can be slow
- Pod being scheduled after PostBind trigger and before NRT update
in those cases DiscardReserved cache will act same as Passthrough cache.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Add new DiscardReservedNodes cache type for NodeResourceTopologyPlugin.
```

### CODE REVIEW

**Comment by PiotrProkop** [comment]
@ffromani @swatisehgal please review.

**Comment by ffromani** [comment]
/ok-to-test
I had offline chat with @PiotrProkop and I'm very in favor of this general direction and open to this kind of additions. However I need to deep dive into the design and the code to see how we can clearly fit this code in the current framework, and unfortunately I won't have the time for that in the next few days.

**Comment by ffromani** [comment]
@PiotrProkop just FYI additionally, I would finish https://github.com/kubernetes-sigs/scheduler-plugins/pull/488 before to tackle this one

**Comment by PiotrProkop** [comment]
> @PiotrProkop just FYI additionally, I would finish #488 before to tackle this one

ok, let me know when #488 will be ready to review.

File: pkg/noderesourcetopology/cache/overreserve.go [hunk 1]
```diff
@@ -70,27 +70,27 @@ func NewOverReserve(lister listerv1alpha1.NodeResourceTopologyLister, indexer No
...
 	defer ov.lock.Unlock()
 	if ov.nodesWithForeignPods.IsSet(nodeName) {
-		return nil, false
+		return nil, fmt.Errorf("foreign pods scheduled on node: %s", nodeName)
 	}
 
 	nrt := ov.nrts.GetNRTCopyByNodeName(nodeName)
 	if nrt == nil {
-		return nil, true
+		return nil, fmt.Errorf("cannot get NodeResourceTopology for node: %s", nodeName)
```
Comment (ffromani) [inline pkg/noderesourcetopology/cache/overreserve.go:82]: This is a trap in the existing code and would have deserved a comment. For historical reasons, NOT having NRT info is a legit case which is meant to be handled as passthrough (no filter). We can revisit this behavior of course, but in another PR :)

**Review by ffromani — COMMENTED** [review]
still need to find time to properly review, but something caught my eye already

**Comment by ffromani** [comment]
@PiotrProkop could you please rebase? I'd be happy to review the week starting on March 6 2023

**Comment by PiotrProkop** [comment]
> @PiotrProkop could you please rebase? I'd be happy to review the week starting on March 6 2023

I am on it!

**Comment by PiotrProkop** [comment]
> > @PiotrProkop could you please rebase? I'd be happy to review the week starting on March 6 2023
> 
> I am on it!

@ffromani  Should be good to review now!

**Comment by ffromani** [comment]
> > > @PiotrProkop could you please rebase? I'd be happy to review the week starting on March 6 2023
> > 
> > 
> > I am on it!
> 
> @ffromani Should be good to review now!

Thanks! I'll start my review later today (March 6)

**Comment by ffromani** [comment]
Sorry for the delay. Though week. Reviewing this PR is still on the top of my TODO.

File: pkg/noderesourcetopology/cache/cache.go [hunk 2]
```diff
@@ -30,7 +30,7 @@ type Interface interface {
 	// The pod argument is used only for logging purposes.
 	// Returns a boolean to signal the caller if the NRT data is clean. If false, then the node has foreign
 	// Pods detected - so it should be ignored or handled differently by the caller.
-	GetCachedNRTCopy(nodeName string, pod *corev1.Pod) (*topologyv1alpha2.NodeResourceTopology, bool)
+	GetCachedNRTCopy(nodeName string, pod *corev1.Pod) (*topologyv1alpha2.NodeResourceTopology, error)
```
Comment (ffromani) [inline pkg/noderesourcetopology/cache/cache.go:33]: this is probably a good idea in general. Ideally, this would be its own commit.

File: pkg/noderesourcetopology/cache/cache.go [hunk 3]
```diff
@@ -56,4 +56,10 @@ type Interface interface {
 
 	// UnreserveNodeResources decrement from the node assumed resources the resources required by the given pod.
 	UnreserveNodeResources(nodeName string, pod *corev1.Pod)
+
+	// PostBind is called after a pod is successfully bound. These plugins are
+	// informational. A common application of this extension point is for cleaning
+	// up. If a plugin needs to clean-up its state after a pod is scheduled and
+	// bound, PostBind is the extension point that it should register.
+	PostBind(nodeName string, pod *corev1.Pod)
```
Comment (ffromani) [inline pkg/noderesourcetopology/cache/cache.go:64]: interesting. I'll review if this can helps the other cache types.

File: pkg/noderesourcetopology/cache/overreserve.go (hunk 1 above)
Comment (ffromani) [reply pkg/noderesourcetopology/cache/overreserve.go:82]: I suggest to turn the error into a log and `return nil, nil` here

File: pkg/noderesourcetopology/cache/passthrough.go [hunk 4]
```diff
@@ -34,17 +36,17 @@ func NewPassthrough(lister listerv1alpha2.NodeResourceTopologyLister) Interface
...
 }
 
-func (pt Passthrough) GetCachedNRTCopy(nodeName string, _ *corev1.Pod) (*topologyv1alpha2.NodeResourceTopology, bool) {
+func (pt Passthrough) GetCachedNRTCopy(nodeName string, _ *corev1.Pod) (*topologyv1alpha2.NodeResourceTopology, error) {
 	klog.V(5).InfoS("Lister for nodeResTopoPlugin", "lister", pt.lister)
 	nrt, err := pt.lister.Get(nodeName)
 	if err != nil {
-		klog.V(5).ErrorS(err, "Cannot get NodeTopologies from NodeResourceTopologyLister")
-		return nil, true
+		return nil, fmt.Errorf("cannot get NodeTopologies from NodeResourceTopologyLister")
```
Comment (ffromani) [inline pkg/noderesourcetopology/cache/passthrough.go:43]: this is another trap, the original code didn't fail. We need to check what happens on the call site with this change, which looks indeed more correct.

File: pkg/noderesourcetopology/filter.go [hunk 5]
```diff
@@ -198,12 +198,10 @@ func (tm *TopologyMatch) Filter(ctx context.Context, cycleState *framework.Cycle
...
 	nodeName := nodeInfo.Node().Name
-	nodeTopology, ok := tm.nrtCache.GetCachedNRTCopy(nodeName, pod)
-	if !ok {
-		return framework.NewStatus(framework.Unschedulable, fmt.Sprintf("invalid node topology data for node %s", nodeName))
-	}
-	if nodeTopology == nil {
-		return nil
+	nodeTopology, err := tm.nrtCache.GetCachedNRTCopy(nodeName, pod)
+	if err != nil {
+		klog.Error(err)
```
Comment (ffromani) [inline pkg/noderesourcetopology/filter.go:203]: maybe `klog.ErrorS(err, ...)` and keep the original log message?

File: pkg/noderesourcetopology/filter.go [hunk 6]
```diff
@@ -198,12 +198,10 @@ func (tm *TopologyMatch) Filter(ctx context.Context, cycleState *framework.Cycle
...
+		return framework.NewStatus(framework.Unschedulable, err.Error())
```
Comment (ffromani) [inline pkg/noderesourcetopology/filter.go:204]: why the other check was removed?

File: pkg/noderesourcetopology/score.go [hunk 7]
```diff
@@ -63,15 +63,10 @@ func (tm *TopologyMatch) Score(ctx context.Context, state *framework.CycleState,
...
-	if !ok {
+	nodeTopology, err := tm.nrtCache.GetCachedNRTCopy(nodeName, pod)
+	if err != nil {
 		klog.V(4).InfoS("noderesourcetopology is not valid for node", "node", nodeName)
-		return 0, nil
-	}
-	if nodeTopology == nil {
-		klog.V(5).InfoS("noderesourcetopology was not found for node", "node", nodeName)
-		return 0, nil
+		return framework.MinNodeScore, nil
```
Comment (ffromani) [inline pkg/noderesourcetopology/score.go:69]: again not sure (literally, need to check) that removing the `if nodeTopology == nil` guard has no ill effects in the flow.

File: apis/config/v1/types.go [hunk 8]
```diff
@@ -149,6 +149,8 @@ type NodeResourceTopologyMatchArgs struct {
 	ScoringStrategy *ScoringStrategy `json:"scoringStrategy,omitempty"`
 	// If > 0, enables the caching facilities of the reserve plugin - which must be enabled
 	CacheResyncPeriodSeconds *int64 `json:"cacheResyncPeriodSeconds,omitempty"`
+	// if set to true, exclude node from scheduling if there are any reserved pods for given node
+	DiscardReservedNodes bool
```
Comment (ffromani) [inline apis/config/v1/types.go:153]: missing JSON annotations. Please make sure the field is fully optional. Do you need to distinguish between "unspecified" and "zero"?

**Review by ffromani — COMMENTED** [review]
initial partial review

File: pkg/noderesourcetopology/cache/passthrough.go (hunk 4 above)
Comment (PiotrProkop) [reply pkg/noderesourcetopology/cache/passthrough.go:43]: this is what it looks like in lister:
```
	obj, exists, err := s.indexer.GetByKey(name)
	if err != nil {
		return nil, err
	}
	if !exists {
		return nil, errors.NewNotFound(v1alpha2.Resource("noderesourcetopology"), name)
	}
```
So if NRT doesn't exist we'll get Not Found error.

File: pkg/noderesourcetopology/filter.go (hunk 6 above)
Comment (PiotrProkop) [reply pkg/noderesourcetopology/filter.go:204]: In my first implementation I removed possibility of returning `nil,nil` from GetCachedNRTCopy, with next patchset I am going to revert it so will add this check again:
```go
if nodeTopology == nil {
	return nil
}
```

File: pkg/noderesourcetopology/filter.go (hunk 5 above)
Comment (PiotrProkop) [reply pkg/noderesourcetopology/filter.go]: :+1:

File: pkg/noderesourcetopology/score.go (hunk 7 above)
Comment (PiotrProkop) [reply pkg/noderesourcetopology/score.go:69]: added this check again

File: apis/config/v1/types.go (hunk 8 above)
Comment (PiotrProkop) [reply apis/config/v1/types.go:153]: I don't need to use `*bool` I can rely on default value of boolean which is `false` 😄

**Comment by PiotrProkop** [comment]
> initial partial review

thanks for the review. I have applied changes from this review.

File: apis/config/v1/types.go (hunk 8 above)
Comment (ffromani) [reply apis/config/v1/types.go:153]: fair enough!

**Comment by ffromani** [comment]
For the record, overall from what I see so far I'm still very much convinced this change is a good direction for the NRT scheduler plugin and that the only issues could be related to properly integrate in the codebase - not about the design, which seems to fit nicely in the current framework and goals. There are already few points that could be generalized to other cache impls to make the codebase better.

File: pkg/noderesourcetopology/cache/discardreserved.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+	//readLock
+	pt.rMutex.RLock()
+	defer pt.rMutex.RUnlock()
+	if t, ok := pt.reservationMap[nodeName]; ok {
+		if len(t) > 0 {
+			return nil, fmt.Errorf("node %s is already reserved by another pod", nodeName)
+		}
+	}
+
+	klog.V(5).InfoS("Lister for nodeResTopoPlugin", "lister", pt.lister)
```
Comment (ffromani) [inline pkg/noderesourcetopology/cache/discardreserved.go:53]: minor: logging the lister is rarely useful in my experience. Perhaps you may want to log some properties of the lister, not the full object?

File: pkg/noderesourcetopology/cache/discardreserved.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+func (pt *DiscardReserved) ReserveNodeResources(nodeName string, pod *corev1.Pod) {
+	klog.V(5).InfoS("nrtcache NRT Reserve", "logID", klog.KObj(pod), "node", nodeName)
+	// write lock
+	pt.rMutex.Lock()
+	defer pt.rMutex.Unlock()
+
+	if pt.reservationMap[nodeName] == nil {
+		pt.reservationMap[nodeName] = make(map[string]bool)
+	}
+	podName := klog.KObj(pod).String()
```
Comment (ffromani) [inline pkg/noderesourcetopology/cache/discardreserved.go:73]: does this guarantee the name is stable and namespace prefixed?

File: pkg/noderesourcetopology/cache/discardreserved.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+	pt.reservationMap[nodeName][podName] = true
```
Comment (ffromani) [inline pkg/noderesourcetopology/cache/discardreserved.go:74]: a design decision we inherited by the podresources api is to use the namespace/name pair to identify the pods.
Changing the podresources API would have been too hard, so this decision leaked all across the stack.
But pod.UID is a stronger identifier and in some case we would have preferred to use it.
So, is the namespace/name pair good enough for you? would using pod.UID lead to more robust/simpler code?

File: pkg/noderesourcetopology/cache/discardreserved.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+	klog.V(5).InfoS("nrtcache NRT", "logID", klog.KObj(pod), "node", nodeName, "reservationMap", pt.reservationMap[nodeName])
```
Comment (ffromani) [inline pkg/noderesourcetopology/cache/discardreserved.go:76]: this easily gets big (thus poorly readable) when the nodes get populated, though. It's fine to log this only at high verbosiness level, but I'm wondering if we can log somehing more useful

File: pkg/noderesourcetopology/cache/discardreserved.go [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+func (pt *DiscardReserved) UnreserveNodeResources(nodeName string, pod *corev1.Pod) {
+	klog.V(5).InfoS("nrtcache NRT Unreserve", "logID", klog.KObj(pod), "node", nodeName)
+	pt.removeReservationForNode(nodeName, pod)
+}
+
+// PostBind is invoked to cleanup reservationMap
+func (pt *DiscardReserved) PostBind(nodeName string, pod *corev1.Pod) {
```
Comment (ffromani) [inline pkg/noderesourcetopology/cache/discardreserved.go:85]: this is interesting. I think I get it but could you please elaborate (gh comment is fine) why you want to remove the reservation on postbind?

**Review by ffromani — COMMENTED** [review]
almost done, need to finish looking at the tests and at the core changes

File: pkg/noderesourcetopology/cache/passthrough.go (hunk 4 above)
Comment (ffromani) [reply pkg/noderesourcetopology/cache/passthrough.go:43]: Uhm, I still tend to believe we DO NOT want to error out here, I recall some past conversations in this area. I'll do dome archeology.

File: pkg/noderesourcetopology/score.go (hunk 7 above)
Comment (ffromani) [reply pkg/noderesourcetopology/score.go:69]: thanks. Could you please add back the log as well?

File: pkg/noderesourcetopology/cache/discardreserved.go (hunk 9 above)
Comment (PiotrProkop) [reply pkg/noderesourcetopology/cache/discardreserved.go:53]: You're right! I need to work on logging part of this PR.

File: pkg/noderesourcetopology/cache/discardreserved.go (hunk 10 above)
Comment (PiotrProkop) [reply pkg/noderesourcetopology/cache/discardreserved.go:73]: Haven't though about it, probably should use pod.UID.

File: pkg/noderesourcetopology/cache/discardreserved.go (hunk 11 above)
Comment (PiotrProkop) [reply pkg/noderesourcetopology/cache/discardreserved.go:74]: pod.UID FTW!

**Review by ffromani — COMMENTED** [review]
Besides a couple items in the initial `GetCachedNRTCopy` PR, the prep work LGTM. I'll finish the review of the core changes later.

File: pkg/noderesourcetopology/score.go (hunk 7 above)
Comment (PiotrProkop) [reply pkg/noderesourcetopology/score.go:69]: :+1:

File: pkg/noderesourcetopology/cache/passthrough.go (hunk 4 above)
Comment (ffromani) [reply pkg/noderesourcetopology/cache/passthrough.go:43]: let's hold on this and the other item regarding GetCachedNRTCopy. I'll do some research and come up with the suggestion. I'm confident postponing these changes should be little hassle.

**Comment by PiotrProkop** [comment]
> Besides a couple items in the initial `GetCachedNRTCopy` PR, the prep work LGTM. I'll finish the review of the core changes later.

Thanks a lot! I'll update PR till EOW.

File: pkg/noderesourcetopology/cache/discardreserved.go (hunk 10 above)
Comment (PiotrProkop) [reply pkg/noderesourcetopology/cache/discardreserved.go:73]: After checking I used this function cause it prefixes pod with namespace but I like using `UID` better, so I am changing implementation to use it right now!

**Comment by PiotrProkop** [comment]
> > Besides a couple items in the initial `GetCachedNRTCopy` PR, the prep work LGTM. I'll finish the review of the core changes later.
> 
> Thanks a lot! I'll update PR till EOW.

PR updated. Decided to go with `pod.UID` as `reservationMap` key. I have also dropped the commit that changed the signature of `GetCachedNRTCopy`.

File: apis/config/v1beta2/types.go [hunk 14]
```diff
@@ -148,6 +148,8 @@ type NodeResourceTopologyMatchArgs struct {
 	// implicitely enables the caching. If zero, disables the caching entirely.
 	// If the cache is enabled, the Reserve plugin must be enabled.
 	CacheResyncPeriodSeconds *int64 `json:"cacheResyncPeriodSeconds,omitempty"`
+	// if set to true, exclude node from scheduling if there are any reserved pods for given node
+	DiscardReservedNodes bool `json:"discardReservedNodes,omitempty"`
```
Comment (ffromani) [inline apis/config/v1beta2/types.go:152]: my only remaining concern is this effectively makes the code ignore `CacheResyncPeriodSeconds` (if given).
I'd start documenting this. We can think about a better UX later on.

File: pkg/noderesourcetopology/cache/discardreserved.go [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+type DiscardReserved struct {
+	rMutex         sync.RWMutex
+	reservationMap map[string]map[types.UID]bool // Key is NodeName, value is Pod UID : reserved status
+	lister         listerv1alpha2.NodeResourceTopologyLister
+}
+
+func NewDiscardReserved(lister listerv1alpha2.NodeResourceTopologyLister) Interface {
+	return &DiscardReserved{
+		lister:         lister,
+		reservationMap: make(map[string]map[types.UID]bool),
```
Comment (ffromani) [inline pkg/noderesourcetopology/cache/discardreserved.go:39]: minor: I wonder if we should use a set here (k8s.io/apimachinery/pkg/util/sets)

File: pkg/noderesourcetopology/cache/discardreserved.go (hunk 13 above)
Comment (ffromani) [reply pkg/noderesourcetopology/cache/discardreserved.go:80]: note really looks very similar (possibly the same) reason we introduced the resync mechanism (using podfingerprint). You can't really be sure about timings here. The NRT update can come anytime. They will come in order, sure, but there's no guarantee about timing and about the sequence of events considering the scheduler event and the NRT update events.

**Review by ffromani — COMMENTED** [review]
thanks for the updates. Codewise LGTM with a note about the config, please check.
I'm OK adding this code to `noderesourcetopology` codebase, it's self-contained and I'm happy to assist your use case, but looks to me you folks are hitting very similar issues which led to the design of the `overreserve` cache, so maybe worth a deeper dive here?

File: pkg/noderesourcetopology/cache/discardreserved.go (hunk 13 above)
Comment (ffromani) [reply pkg/noderesourcetopology/cache/discardreserved.go:80]: Unless I'm missing something, you are totally correct that a NRT update is expected once `PostBind` triggers, but there's no guarantee about WHEN this will happen. NFD can be crashlooping, network can be slow, we can have any network glitch which prevents the NRT update to be propagated [temporary glitch]. But most notably there's no guarantee with respect the occurrence of new pods to be scheduled.

In yet another different way (but still related): what ensures that a NRT update is received before a new scheduling attempt is performed against this node when a new pod comes (new pod scheduling is of course completely async).

In the case of the overreserve, we have the podfingerprint machinery and we intentionally are using a pessimistic approach.

Comment (PiotrProkop) [reply pkg/noderesourcetopology/cache/discardreserved.go:80]: You're right but `overreserve` doesn't cache anything if a POD requests resources from multiple NUMA domains(thanks to pesimistic approach and subtracting from every NUMA domain). And it will fetch data from NRT anyway (and temporary glitches also affects this 😄 ). The idea behind Locking nodes is to minimize Admission Errors not to eliminate them.

**Comment by ffromani** [comment]
(copypasting here to ensure visibility)

We covered all the facets I can think of. I'll do a last pass for due diligence to integration tests, but we are almost there. We need two things more, both should be quick and easy:

-  please document as comment in the config that enabling DiscardReservedNodes takes precedence over CacheResyncPeriodSeconds. A code comment is fine.
-  please describe (for historical record) the usecase and the design/conceptual differences between this cache impl and overreserve in the PR cover letter and/or in `cache/discardreserved.go` (preferably both, just copypaste the same block of text). 

I think we already discussed the rationale for this cache impl, it would be fine to just summarize our past conversation and copy/paste it as comment so we have the record for the future in a easy to find place.

**Comment by PiotrProkop** [comment]
> (copypasting here to ensure visibility)
> 
> We covered all the facets I can think of. I'll do a last pass for due diligence to integration tests, but we are almost there. We need two things more, both should be quick and easy:
> 
>     * please document as comment in the config that enabling DiscardReservedNodes takes precedence over CacheResyncPeriodSeconds. A code comment is fine.
> 
>     * please describe (for historical record) the usecase and the design/conceptual differences between this cache impl and overreserve in the PR cover letter and/or in `cache/discardreserved.go` (preferably both, just copypaste the same block of text).
> 
> 
> I think we already discussed the rationale for this cache impl, it would be fine to just summarize our past conversation and copy/paste it as comment so we have the record for the future in a easy to find place.

Thanks for the review!

I have added comments in the code and updated PR description to explain rationale.

File: pkg/noderesourcetopology/cache/discardreserved_test.go [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	nrtCache.ReserveNodeResources("node1", &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pod",
+			Namespace: "test",
+			UID:       "some-uid",
+		},
+	})
+	nodePods, ok := nrtCache.reservationMap["node1"]
```
Comment (ffromani) [inline pkg/noderesourcetopology/cache/discardreserved_test.go:112]: let's improve the tests later on
1. adding (more) integration tests
2. avoiding to access private fields/methods

**Review by ffromani — COMMENTED** [review]
/approve
/lgtm

self-contained change. Useful to explore new directions which may benefit `overreserve in the future, so win-win
